allennlp/commands/build_vocab.py

```diff
@@ -0,0 +1,100 @@
+"""
+Subcommand for building a vocabulary from a training config.
+"""
+
+import argparse
+import json
+import logging
+import os
+import tarfile
+import tempfile
+
+from overrides import overrides
+
+from allennlp.commands.subcommand import Subcommand
+from allennlp.common.params import Params
+from allennlp.training.util import make_vocab_from_params
+
+
+logger = logging.getLogger(__name__)
+
+
+@Subcommand.register("build-vocab")
+class BuildVocab(Subcommand):
+    @overrides
+    def add_subparser(self, parser: argparse._SubParsersAction) -> argparse.ArgumentParser:
+        description = """Build a vocabulary from an experiment config file."""
+        subparser = parser.add_parser(self.name, description=description, help=description)
+
+        subparser.add_argument("param_path", type=str, help="path to an experiment config file")
+
+        subparser.add_argument(
+            "output_path", type=str, help="path to save the vocab tar.gz file to"
+        )
+
+        subparser.add_argument(
+            "-f",
+            "--force",
+            action="store_true",
+            help="force write if the output_path already exists",
+        )
+
+        subparser.add_argument(
+            "-o",
+            "--overrides",
+            type=str,
+            default="",
+            help=(
+                "a json(net) structure used to override the experiment configuration, e.g., "
+                "'{\"iterator.batch_size\": 16}'.  Nested parameters can be specified either"
+                " with nested dictionaries or with dot syntax."
+            ),
+        )
+
+        subparser.set_defaults(func=build_vocab_from_args)
+
+        return subparser
+
+
+def build_vocab_from_args(args: argparse.Namespace):
+    if not args.output_path.endswith(".tar.gz"):
+        raise ValueError("param 'output_path' should end with '.tar.gz'")
+
+    if os.path.exists(args.output_path) and not args.force:
+        raise RuntimeError(f"{args.output_path} already exists. Use --force to overwrite.")
+
+    output_directory = os.path.dirname(args.output_path)
+    os.makedirs(output_directory, exist_ok=True)
+
+    params = Params.from_file(args.param_path)
+
+    with tempfile.TemporaryDirectory() as temp_dir:
+        # Serializes the vocab to 'tempdir/vocabulary'.
+        make_vocab_from_params(params, temp_dir)
+
+        # We create a temp file to archive the vocab to in case something goes wrong
+        # while creating the archive so we're not left with a corrupted file.
+        temp_archive = tempfile.NamedTemporaryFile(
```

Comment:
This whole "create inside a temp file" seems like a complex piece of code that happens in a few places and could maybe be shared.  Then again, you've got some vocabulary-specific stuff in here.  What happens in other places when we save the vocab?  Is there similar logic there?  Just a note to consider whether it's worth sharing this logic somewhere.  Might not be worth it.

Reply:
Ah, good point. I forgot we have the `CacheFile` context manager that already handles this logic.